### DISCUSSION TITLE
Adding a custom logger

### DISCUSSION THREAD

**Discussion by ripvannwinkler** [medusajs/medusa#4708] [Q&A]
For a recent project we launched with Medusa, we had a need to add an additional Winston transport to send the logs to a log collector we use internally. I had to stumble through the source code to figure out how to do this. It works fine, but I would suggest getting the docs updated to support this scenario. I'm willing to submit a PR for updating the docs, but I'm not sure where it would be relevant to add. Thoughts?

Below is an example of the loader we used to add our external log collector:

```typescript
import { SeqTransport } from '@datalust/winston-seq';
import type { Logger } from '@medusajs/medusa';
import type { AwilixContainer } from 'awilix';
import winston from 'winston';

function createFileTransport() {
	const level = 'debug';
	const maxFiles = 20;
	const maxsize = 10485760; // 10MB
	const filename = process.env.APP_LOG_FILE || 'app.log';

	return new winston.transports.File({
		level,
		filename,
		maxsize,
		maxFiles,
	});
}

function createSeqTransport() {
	return new SeqTransport({
		serverUrl: '*********************************',
		apiKey: '*****************************',
		onError: (e) => {
			console.error(e);
		},
	});
}

function addCustomLoggers(loggerInstance: winston.Logger) {
	const appName = process.env.MEDUSA_APP_NAME || 'My App Name';
	const appEnv = process.env.APP_ENVIRONMENT || 'Production';

	try {
		loggerInstance.configure({
			defaultMeta: {
				...loggerInstance.defaultMeta,
				Application: appName,
				Environment: appEnv,
			},
			transports: [
				...loggerInstance.transports, //
				createFileTransport(),
				createSeqTransport(),
			],
		});
	} catch (e) {
		console.error(`Error adding loggers: ${e}`);
	}
}

export default async function (container: AwilixContainer): Promise<void> {
	const logger = container.resolve<Logger>('logger');
	const loggerInstance = (logger as any).loggerInstance_ as winston.Logger;
	addCustomLoggers(loggerInstance);
}
```

As an aside, the reporter class that is registered as the logger has a loggerInstance_ field that is required to enable the above, but it's not exposed via the type declaration. It would have been helpful to me if this had been the case.

**Comment by pevey** [comment]
Thanks for sharing this.  On your last comment, I think you are referring to this?
```
export class Reporter {
  protected activities_: Record<string, any>
  protected loggerInstance_: winston.Logger
  protected ora_: typeof ora

  constructor({ logger, activityLogger }) {
    this.activities_ = {}
    this.loggerInstance_ = logger
    this.ora_ = activityLogger
  }
...
  ```
  If I understand correctly, you are suggesting that the protected property loggerInstance_ be changed to public, so that it can be referenced directly and modified.  This seems like a good idea.

**Reply by ripvannwinkler** [reply]
Correct. I think the loggerInstance_ should be public (or exposed via a public getter) and documentation should be added on the recommended way to customize. My sample works, but if there's a better way, I'd be open to it.
